modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuilder.java

```diff
@@ -0,0 +1,220 @@
+                new LinkedBlockingQueue<>(),
+                NamedThreadFactory.create(nodeName, "build-index", LOG)
+        );
+    }
+
+    /**
+     * Stops the index builder.
+     */
+    void stop() {
+        shutdownAndAwaitTermination(buildIndexExecutor, 10, TimeUnit.SECONDS);
+    }
+
+    /**
+     * Initializes the build of the index.
+     */
+    void startIndexBuild(TableIndexView tableIndexView, TableImpl table) {
+        for (int partitionId = 0; partitionId < table.internalTable().partitions(); partitionId++) {
+            buildIndexExecutor.submit(new BuildIndexTask(table, tableIndexView, partitionId, true));
+        }
+    }
+
+    /**
+     * Task of building a table index for a partition.
+     *
+     * <p>Only the leader of the raft group will manage the building of the index. Leader sends batches of row IDs via
+     * {@link BuildIndexCommand}, the next batch will only be send after the previous batch has been processed.
+     *
+     * <p>Index building itself occurs locally on each node of the raft group when processing {@link BuildIndexCommand}. This ensures that
+     * the index build process in the raft group is consistent and that the index build process is restored after restarting the raft group
+     * (not from the beginning).
+     */
+    private class BuildIndexTask implements Runnable {
+        private final TableImpl table;
+
+        private final TableIndexView tableIndexView;
+
+        private final int partitionId;
+
+        private final boolean firstBatch;
+
+        private BuildIndexTask(TableImpl table, TableIndexView tableIndexView, int partitionId, boolean firstBatch) {
+            this.table = table;
+            this.tableIndexView = tableIndexView;
+            this.partitionId = partitionId;
+            this.firstBatch = firstBatch;
+        }
+
+        @Override
+        public void run() {
+            if (!busyLock.enterBusy()) {
+                return;
+            }
+
+            try {
+                InternalTable internalTable = table.internalTable();
+
+                RaftGroupService raftGroupService = internalTable.partitionRaftGroupService(partitionId);
+
+                if (!isLocalNodeLeader(raftGroupService)) {
+                    // TODO: IGNITE-19053 Must handle the change of leader
+                    return;
+                }
+
+                RowId lastBuildRowId = internalTable.storage().getOrCreateIndex(partitionId, tableIndexView.id()).getLastBuildRowId();
+
+                if (lastBuildRowId == null) {
+                    // Index has already been built.
+                    return;
+                }
+
+                if (firstBatch) {
+                    LOG.info("Start building the index: [{}]", createCommonTableIndexInfo());
+                }
+
+                List<RowId> batchRowIds = createBatchRowIds(lastBuildRowId, BUILD_INDEX_ROW_ID_BATCH_SIZE);
+
+                boolean finish = batchRowIds.size() < BUILD_INDEX_ROW_ID_BATCH_SIZE;
```

Comment:
But this will make it possible to coordinate the indices on all members of the raft group, is that bad? 
no additional steps are needed to handle the situation of creating indexes for an empty partition on all members of raft group.